external/storm-kinesis/src/main/java/org/apache/storm/kinesis/spout/KinesisRecordsManager.java

```diff
@@ -0,0 +1,566 @@
+        }
+        path += config.getZkInfo().getZkNode();
+        if (!config.getZkInfo().getZkNode().endsWith("/")) {
+            path += "/";
+        }
+        path += (config.getStreamName() + "/" + shardId);
+        return path;
+    }
+
+    private void commitState (String path, Map<Object, Object> state) {
+        byte[] bytes = JSONValue.toJSONString(state).getBytes(Charset.forName("UTF-8"));
+        try {
+            if (curatorFramework.checkExists().forPath(path) == null) {
+                curatorFramework.create()
+                        .creatingParentsIfNeeded()
+                        .withMode(CreateMode.PERSISTENT)
+                        .forPath(path, bytes);
+            } else {
+                curatorFramework.setData().forPath(path, bytes);
+            }
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private Map<Object, Object> readState (String path) {
+        try {
+            Map<Object, Object> state = null;
+            byte[] b = null;
+            if (curatorFramework.checkExists().forPath(path) != null) {
+                b = curatorFramework.getData().forPath(path);
+            }
+            if (b != null) {
+                state = (Map<Object, Object>) JSONValue.parse(new String(b, "UTF-8"));
+            }
+            return state;
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    // fetch records from kinesis starting at sequence number for message passed as argument. Any other messages fetched and are in the failed queue will also
+    // be kept in memory to avoid going to kinesis again for retry
+    private void fetchFailedRecords (KinesisMessageId kinesisMessageId) {
+        // if shard iterator not present for this message, get it
+        if (!shardIteratorPerFailedMessage.containsKey(kinesisMessageId)) {
+            refreshShardIteratorForFailedRecord(kinesisMessageId);
+        }
+        String shardIterator = shardIteratorPerFailedMessage.get(kinesisMessageId);
+        LOG.debug("Fetching failed records for shard id :" + kinesisMessageId.getShardId() + " at sequence number " + kinesisMessageId.getSequenceNumber() +
+                " using shardIterator " + shardIterator);
+        try {
+            GetRecordsResult getRecordsResult = fetchRecords(shardIterator);
+            if (getRecordsResult != null) {
+                List<Record> records = getRecordsResult.getRecords();
+                LOG.debug("Records size from fetchFailedRecords is " + records.size());
+                // update the shard iterator to next one in case this fetch does not give the message.
+                shardIteratorPerFailedMessage.put(kinesisMessageId, getRecordsResult.getNextShardIterator());
+                if (records.size() == 0) {
+                    LOG.debug("No records returned from kinesis. Hence sleeping for 1 second");
+                    Thread.sleep(1000);
```

Comment:
I would prefer this to be not configurable. Reason is kinesis api somewhere mentioned in their documentation that if you dont receive any records it is good to sleep for a second. Reason is that they some limits for GetRecords request per second and it will throw an exception if that limit is reached or violated. Although I am catching that exception I wanted to avoid that. Making it configurable and go beyond 1 second can slow down the spout in some cases as well. For example, sometimes it takes multiple requests to get to the portion of the shard where the message at a sequence number you want to read is present. In that case it does not make sense to backoff in an exponential fashion. You just want to make sure you get there sooner without violating the limit.